### ISSUE TITLE
Optional model training 🍭

### ISSUE THREAD

**Issue by KarelZe** [KarelZe/thesis#97]
- [ ] Calculate cardinality dynamiccally in Classifier
- [ ] Improve gpu utilization of `TabTransformer`
- [ ] Try out moving window approach e. g., for CatBoost
- [ ] Get PyTorch 2.0 running once stable. See [here.](https://pytorch.org/get-started/pytorch-2.0/)
- [ ] Look into callbacks of PyTorch Lightning / huggingface transformers
- [ ] Set-up PyTorch profiler as shown [here](https://www.youtube.com/watch?v=5RUOrXl3nag)
- [ ] Try out other samplers beside `TPESampler` like `RandomSampler`. See how it affects the results
- [ ] Look into implementing transformers using [xformers](https://github.com/facebookresearch/xformers)
- [ ] Let `torch.tensors` and `np.array` share memory. Could be useful for pre-training. See [here.](https://pytorch.org/docs/stable/generated/torch.Tensor.numpy.html)
 - [ ] Add decay hyperparameter for gradient boosting / transformer
- [ ] Run feature selection in CatBoost. See [here.](https://catboost.ai/en/docs/concepts/python-reference_catboost_select_features). Sample could be fine.
- [ ] Try out target enconding
- [ ] Add frequency encoding
```python
# found at: https://www.kaggle.com/competitions/ieee-fraud-detection/discussion/108575
temp = df['card1'].value_counts().to_dict()
df['card1_counts'] = df['card1'].map(temp)
```
- [ ] Study feature interactions:
```python
feature_interaction = [[X.columns[interaction[0]], X.columns[interaction[1]], interaction[2]] for i,interaction in interactions.iterrows()]
feature_interaction_df = pd.DataFrame(feature_interaction, columns=['feature1', 'feature2', 'interaction_strength'])
feature_interaction_df.head(10)
```
- [ ] Compare different feature scaling e. g., normalization, z-score-normalization, robust scaler, quantile transformer etc. approaches for neural net. See [here.](https://machinelearningmastery.com/robust-scaler-transforms-for-machine-learning/) and [here.](https://github.com/KarelZe/thesis/blob/main/notebooks/3.0b-mb-comparsion-transformations.ipynb)
- [ ] Study effects of quantization. How can one assist quantization with feature engineering? For technical background see [here.](https://developer.nvidia.com/blog/gradient-boosting-decision-trees-xgboost-cuda/) and [here.](https://catboost.ai/en/docs/concepts/python-reference_save_quantization_borders)
- [ ] Try out nicer progress bar. See [here.](https://rich.readthedocs.io/en/stable/progress.html)
